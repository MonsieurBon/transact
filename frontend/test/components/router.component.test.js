import {
  fixture,
  fixtureCleanup,
} from '@open-wc/testing-helpers/index-no-side-effects.js';
import Router from '../../src/components/router.component.js';

describe('Router component', () => {
  let router;

  beforeAll(() => {
    router = {
      register: jest.fn(),
      deregister: jest.fn(),
    };
    window.customElements.define('t-router', Router(router));
  });

  afterEach(() => {
    fixtureCleanup();
    jest.clearAllMocks();
  });

  it('should render', async () => {
    const el = await fixture('<t-router></t-router>');
    expect(el.shadowRoot.innerHTML).toMatchSnapshot();
  });

  it('should register itself with the router', async () => {
    const el = await fixture('<t-router></t-router>');
    expect(router.register).toHaveBeenCalledTimes(1);
    expect(router.register).toHaveBeenCalledWith(el);
  });

  it('should deregister itself from the router', async () => {
    const el = await fixture('<t-router></t-router>');
    el.parentElement.removeChild(el);

    expect(router.register).toHaveBeenCalledTimes(1);
    expect(router.deregister).toHaveBeenCalledTimes(1);
    expect(router.deregister).toHaveBeenCalledWith(el);
  });

  it('should render the current path', async () => {
    history.pushState({}, '', '/foo');
    const el = await fixture(
      '<t-router><div data-route="/foo"></div><div data-route="/bar"></div></t-router>'
    );

    expect(el.shadowRoot.innerHTML).toMatchSnapshot();
  });

  it('should replace the rendered node', async () => {
    history.pushState({}, '', '/foo');
    const el = await fixture(
      '<t-router><div data-route="/foo"></div><div data-route="/bar"></div></t-router>'
    );

    expect(el.shadowRoot.innerHTML).toMatchSnapshot();

    history.pushState({}, '', '/bar');
    el.navigationStateChanged();

    expect(el.shadowRoot.innerHTML).toMatchSnapshot();
  });

  it('should not rerender if not necessary', async () => {
    history.pushState({}, '', '/foo');
    const el = await fixture(
      '<t-router><div data-route="/foo"></div><div data-route="/bar"></div></t-router>'
    );

    expect(el.navigationStateChanged()).toBeFalsy();
  });
});
