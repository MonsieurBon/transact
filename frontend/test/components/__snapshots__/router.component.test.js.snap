// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router component should render 1`] = `
"
<style>
  :host {
    contain: content;
  }
  
  slot {
    display: none;
  }
</style>
<div id=\\"router-outlet\\"></div>
<slot>
</slot>"
`;

exports[`Router component should render the current path 1`] = `
"
<style>
  :host {
    contain: content;
  }
  
  slot {
    display: none;
  }
</style>
<div id=\\"router-outlet\\"><div data-route=\\"/foo\\"></div></div>
<slot>
</slot>"
`;

exports[`Router component should replace the rendered node 1`] = `
"
<style>
  :host {
    contain: content;
  }
  
  slot {
    display: none;
  }
</style>
<div id=\\"router-outlet\\"><div data-route=\\"/foo\\"></div></div>
<slot>
</slot>"
`;

exports[`Router component should replace the rendered node 2`] = `
"
<style>
  :host {
    contain: content;
  }
  
  slot {
    display: none;
  }
</style>
<div id=\\"router-outlet\\"><div data-route=\\"/bar\\"></div></div>
<slot>
</slot>"
`;
